applications/FluidDynamicsApplication/custom_processes/mass_conservation_check_process.cpp

```diff
@@ -0,0 +1,265 @@
+            // writing an output at a given frequncy
+            this->ComputeVolumesOfFluids( posVol, negVol );
+            std::cout << " --- Volume Checking Process --- " << std::endl;
+            std::cout << "   positive Volume = " << posVol << std::endl;
+            if ( mCompareToInitial ){ 
+                std::cout << "   ( " << posVol / mInitialPositiveVolume * 100.0 << "% of initial )" << std::endl; 
+            }
+            std::cout << "   negative Volume = " << negVol << std::endl;
+            if ( mCompareToInitial ){ 
+                std::cout << "   ( " << negVol / mInitialNegativeVolume * 100.0 << "% of initial )" << std::endl; 
+            }
+            std::cout << " --- --- --- --- --- --- --- --- " << std::endl;
+
+            if (mWriteToLogFile){
+                std::ofstream myfile;
+                myfile.open ("volumeLog.txt", std::ios::out | std::ios::app);
+                if ( myfile.is_open() ){
+
+                    myfile << mCounter << "    ";
+                    myfile << posVol << "    ";
+                    if ( mCompareToInitial ){ 
+                        myfile << posVol / mInitialPositiveVolume * 100.0 << "    "; 
+                    }
+                    myfile << negVol << "    ";
+                    if ( mCompareToInitial ){ 
+                        myfile << negVol / mInitialNegativeVolume * 100.0 << "\n"; 
+                    } 
+                    myfile.close();
+                }
+            }
+        }
+
+        mCounter++;
+    } else {
+        return;
+    }
+};
+
+
+/* Private functions ****************************************************/
+
+void MassConservationCheckProcess::ComputeVolumesOfFluids( double &positiveVolume, double &negativeVolume )
+{
+    // useless containers
+    Matrix rShapeFunctionsPos, rShapeFunctionsNeg;
+    GeometryType::ShapeFunctionsGradientsType rShapeDerivativesPos, rShapeDerivativesNeg;
+    Vector null3(3, 0.0);
+    Vector null4(4, 0.0);
+
+    // abbreviation of variable names
+    double Vol;
+    double& posVol = positiveVolume;
+    double& negVol = negativeVolume;
+
+    // initalisation
+    posVol = 0.0;
+    negVol = 0.0;
+
+    for (int i_elem = 0; i_elem < static_cast<int>(mrModelPart.NumberOfElements()); ++i_elem){
+        // iteration over all elements
+        auto it_elem = mrModelPart.ElementsBegin() + i_elem;
+        
+        auto p_geom = it_elem->pGetGeometry();
+        unsigned int ptCountPos = 0;
+        unsigned int ptCountNeg = 0;
+
+        // instead of using data.isCut()
+        for (unsigned int pt = 0; pt < p_geom->Points().size(); pt++){
+            if ( p_geom->GetPoint(pt).FastGetSolutionStepValue(DISTANCE) > 0.0 ){
+                ptCountPos++;
+            } else {
+                ptCountNeg++;
+            }
+        }
+
+        if ( ptCountPos == p_geom->Points().size() ){
+            // all nodes are positive
+            if( p_geom->Points().size() == 3 ){
```

Comment:
`DomainSize()` geometries method computes the area for 2D geometries and the volume for 3D ones, so I guess that you can avoid this if statement.

Reply:
Thank you. That makes it simpler.